File: cses/07_mathematics/2164_josephus-queries/solution.py
import sys


def solve(n, k):
    # Trả về số hiệu (1-indexed) của đứa trẻ bị loại ở lượt thứ k.
    # Thu nhỏ bài toán theo tầng: trong vòng quét đầu các vị trí chẵn
    # 2,4,...,2*floor(n/2) bị loại. Nếu k rơi vào vòng đầu, đáp án là 2*k.
    # Ngược lại chỉ còn các vị trí lẻ sống sót nên quy về vòng tròn nhỏ hơn,
    # giữ phép biến đổi tuyến tính answer = A + B * (đáp án bài toán con).
    A = 0
    B = 1
    while True:
        m = n // 2                      # số lần loại trong vòng quét đầu
        if k <= m:                      # lượt loại thứ k rơi vào vòng đầu
            return A + B * (2 * k)
        r = k - m                       # chỉ số loại trong vòng nhỏ (các vị trí lẻ)
        if n % 2 == 0:
            # n chẵn: còn lại 1,3,...,n-1 (m đứa), tiếp theo loại 3 => bài toán gốc
            # trên m đứa. Ánh xạ original_pos = 2*sub - 1  ->  A -= B, B *= 2.
            A -= B
            B *= 2
            n = m
            k = r
        else:
            # n lẻ: còn lại 1,3,...,n (m+1 đứa), số 1 bị loại ngay sau vòng đầu.
            if r == 1:
                return A + B            # số 1 (vị trí gốc = 2*1-1 = 1)
            # Sau khi loại số 1, phần còn lại 3,5,...,n thành bài toán gốc trên m đứa.
            # Ánh xạ original_pos = 2*(1 + sub) - 1 = 1 + 2*sub  ->  A += B, B *= 2.
            A += B
            B *= 2
            n = m
            k = r - 1


def main():
    data = sys.stdin.buffer.read().split()
    q = int(data[0])
    out = []
    idx = 1
    for _ in range(q):
        n = int(data[idx]); k = int(data[idx + 1]); idx += 2
        out.append(str(solve(n, k)))
    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


main()

File: cses/07_mathematics/2164_josephus-queries/solution.cpp
#include <iostream>

// Trả về số hiệu (1-indexed) của đứa trẻ bị loại ở lượt thứ k trên vòng tròn n đứa.
// Khử đệ quy theo tầng: mỗi tầng thu vòng tròn về nửa kích thước, đồng thời giữ phép
// biến đổi tuyến tính answer = offset + scale * sublabel để ánh xạ ngược về nhãn gốc.
long long kth_removed(long long n, long long k) {
    long long offset = 0;
    long long scale = 1;

    while (true) {
        // m = số đứa bị loại trong vòng quét đầu (các vị trí chẵn).
        const long long first_round = n / 2;
        // Nếu k rơi vào vòng quét đầu, đáp án là vị trí chẵn thứ k => 2*k.
        if (k <= first_round) {
            return offset + scale * (2 * k);
        }

        // r = chỉ số loại trong vòng nhỏ (chỉ còn các vị trí lẻ sống sót).
        const long long remaining_k = k - first_round;
        if ((n & 1LL) == 0) {
            // n chẵn: các đứa còn lại 1,3,... ánh xạ sublabel x thành 2*x-1.
            offset -= scale;
            scale *= 2;
            n = first_round;
            k = remaining_k;
        } else {
            // n lẻ: sau khi loại hết vị trí chẵn thì số 1 bị loại ngay tiếp theo.
            if (remaining_k == 1) {
                return offset + scale;
            }
            // Phần còn lại 3,5,... ánh xạ sublabel x thành 2*x+1.
            offset += scale;
            scale *= 2;
            n = first_round;
            k = remaining_k - 1;
        }
    }
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int queries;
    std::cin >> queries;
    while (queries-- > 0) {
        long long n, k;
        std::cin >> n >> k;
        std::cout << kth_removed(n, k) << '\n';
    }
    return 0;
}
